Add LoongArch SIMD intrinsics

heiher opened the pull request:
This PR adds LoongArch SIMD (LSX/LASX) intrinsics and testcases.

Amanieu reviewed (commented):
Are the tests also automatically generated, or just the intrinsic definitions?

heiher commented on crates/stdarch-gen/src/main.rs:
```diff
@@ -3183,6 +3186,10 @@ fn get_call(
     let args: Vec<String> = env::args().collect();
     let in_file = args.get(1).cloned().unwrap_or_else(|| IN.to_string());
+    let arch_out = args.get(2).cloned().unwrap_or_else(|| OUT.to_string());
+    if arch_out == "loongarch64" {
```
Okay. I'll named the generator `stdarch-gen-${cpuarch}`:

* stdarch-gen-arm (neon)
* stdarch-gen-loongarch (lsx/lasx)

Thanks.

heiher commented:
> Are the tests also automatically generated, or just the intrinsic definitions?

The tests are also automatically generated, but not by this generator.

Amanieu commented:
Can you also include the generator used for making the tests? I don't want to include any auto-generated code in this repo without also including the generator used to create it.

Finally, can you also commit a copy of the header files used to generate the .spec files?

heiher commented:
> Can you also include the generator used for making the tests? I don't want to include any auto-generated code in this repo without also including the generator used to create it.
> 
> Finally, can you also commit a copy of the header files used to generate the .spec files?

Thanks for your review. I have updated the patch to include code for generating test cases and also added copying of header files.

Amanieu commented on crates/core_arch/src/loongarch64/lasx/mod.rs:
```diff
@@ -0,0 +1,16 @@
+pub use self::generated::*;
+
+#[rustfmt::skip]
+mod tests;
```
Move the `#[cfg(test)]` here so it doesn't need to be in the generated code.

Amanieu commented on crates/stdarch-gen-loongarch/README.md:
```diff
@@ -0,0 +1,35 @@
+# Generate tests
+OUT_DIR=`pwd`/crates/stdarch-gen-loongarch cargo run -p stdarch-gen-loongarch -- crates/stdarch-gen-loongarch/lsx.spec test
+# On LoongArch
+gcc -o lsx crates/stdarch-gen-loongarch/lsx.c -mlasx
```
Can you change these instructions to use qemu-loongarch64 and a cross-compiler (loongarch64-linux-gnu-gcc)? This makes it easier for people to run without access to loongarch hardware.

heiher replied:
I tested on [this git commit](https://github.com/qemu/qemu/commit/dccbaf0cc0f1744ffd7562a3dc60e4fc99fd9d44) of QEMU using qemu-loongarch64 and a cross-compiler (loongarch64-unknown-linux-gnu-gcc), and I was able to generate the correct test cases. Thanks.

Amanieu reviewed (approved):
This looks great! Just a few minor changes and this is good to merge.

heiher commented on crates/stdarch-gen-loongarch/README.md:
```diff
@@ -0,0 +1,35 @@
+
+# Generate tests
+OUT_DIR=`pwd`/crates/stdarch-gen-loongarch cargo run -p stdarch-gen-loongarch -- crates/stdarch-gen-loongarch/lsx.spec test
+# On LoongArch
```
I'll remove this line.

Amanieu commented on crates/core_arch/src/loongarch64/lasx/tests.rs:
```diff
@@ -0,0 +1,14692 @@
+// This code is automatically generated. DO NOT MODIFY.
+// See crates/stdarch-gen-loongarch/README.md
+
+mod tests {
```
This module is not needed, it is already inside a module called `tests`.

heiher replied:
Done.
